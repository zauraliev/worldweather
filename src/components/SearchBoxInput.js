import React from "react";

const SearchBoxInput = props => (
  <div className="form-group">
    <input
      type="text"
      name={props.inputName}
      className="form-control"
      id={props.inputId}
      placeholder={props.placeholder}
      onChange={props.updateFormInputs}
      onBlur={props.updateFormInputs}
      required={props.errorType}
    />
    {props.errorType ? (
      <div className="invalid-feedback">Please enter {props.placeholder} field</div>
    ) : null}
  </div>
);

export default SearchBoxInput;
