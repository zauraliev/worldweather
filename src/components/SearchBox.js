import React, { Component } from "react";
import SearchBoxInput from "./SearchBoxInput";

//import Autocomplete from "react-autocomplete";
//let selectedValue = "";

class SearchBox extends Component {
  getWeather = async e => {
    e.preventDefault();
    const api_key = "01f0312628c56d7fcadfdee00da7c7a3";
    let api_base_url = "https://api.openweathermap.org/data/2.5/weather?q=";
    let city = e.target.elements.city.value;
    let country = e.target.elements.country.value;

    if (city && country) {
      try {
        this.props.updateState({
          isLoading: true
        });
        const callAPI = await fetch(`${api_base_url}${city},${country}&appid=${api_key}`);

        const data = await callAPI.json();

        this.props.updateState({
          location: `${data.name} City, ${data.sys.country}`,
          temperature: this.props.calculateTemp(data.main.temp),
          humidity: `${data.main.humidity}%`,
          conditions: this.props.capitalize(data.weather[0].description),
          error: "",
          city: `${city}`,
          country: `${country}`,
          showResult: true,
          isLoading: false
        });
      } catch (error) {
        this.props.updateState({
          error: error.message,
          isLoading: false
        });
      }
    } else {
      this.props.updateState({
        touched: {
          city: true,
          country: true
        }
      });
    }
  };

  updateFormInputs = e => {
    e.preventDefault();
    this.props.updateState({ [e.target.name]: e.target.value });
    if (e.target.name === "city") {
      this.props.updateState({
        city: e.target.value,
        touched: {
          city: true,
          country: this.props.touched.country
        }
      });
    }
    if (e.target.name === "country") {
      this.props.updateState({
        country: e.target.value,
        touched: {
          city: this.props.touched.city,
          country: true
        },
        countryCode: e.target.value
      });
    }
  };
  render() {
    const cityError = !this.props.city && this.props.touched.city;
    const countryError = !this.props.country && this.props.touched.country;
    const errorCondition = cityError || countryError;
    const validateFormError = errorCondition ? "was-validated" : "";
    return (
      <form
        id="weatherForm"
        onSubmit={this.getWeather}
        className={"needs-validation " + validateFormError}
        noValidate
      >
        <SearchBoxInput
          inputName="city"
          inputId="city"
          placeholder="City"
          updateFormInputs={this.updateFormInputs}
          errorType={cityError}
        />
        <SearchBoxInput
          inputName="country"
          inputId="country"
          placeholder="Country"
          updateFormInputs={this.updateFormInputs}
          errorType={countryError}
        />
        <div className="form-group">
          <button className="btn btn-main btn-md btn-block">
            <span className="btn-text">Get Weather</span>
          </button>
        </div>
      </form>
    );
  }
}

export default SearchBox;
