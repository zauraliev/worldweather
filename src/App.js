import React, { Component } from "react";
import Header from "./components/Header";
import SearchBox from "./components/SearchBox";
import SearchResult from "./components/SearchResult";

class App extends Component {
  state = {
    location: "",
    temperature: "",
    humidity: "",
    conditions: "",
    touched: {
      city: false,
      country: false
    },
    city: "",
    country: "",
    showResult: false,
    isLoading: false,
    value: "",
    autocompleteData: []
  };
  updateState = state => {
    this.setState(state);
  };
  capitalize = ([first, ...rest]) => first.toUpperCase() + rest.join("").toLowerCase();

  calculateTemp = temp => {
    const tempSign = "Â°F";
    const dayTimeTemp = Math.round(1.8 * (temp - 273.15) + 32);

    return `${dayTimeTemp}${tempSign}`;
  };

  componentDidMount = async () => {
    // This part is for Autocomplete inputs for SearchBox
    // will continue work on it later
    const countriesData = require("./api/countries");
    const citiesData = require("./api/country_codes_cities");

    const countriesWithCodes = [];

    if (countriesData) {
      for (let i = 0; i < countriesData.length; i++) {
        let obj = {};
        obj.label = countriesData[i].name;
        obj.value = countriesData[i].alpha2Code;
        countriesWithCodes.push(obj);
      }

      var citiesClone = JSON.parse(JSON.stringify(citiesData));

      this.setState({
        countries: countriesWithCodes,
        autocompleteData: countriesWithCodes,
        cities: citiesClone
      });
    }
  };

  render() {
    return (
      <div className="wrapper">
        <div className="main">
          <div className="container-fluid">
            <div className="row justify-content-center align-items-center">
              <div className="col-4">
                <div className="form-group">
                  <Header />
                </div>
                <SearchBox
                  city={this.state.city}
                  country={this.state.country}
                  error={this.state.error}
                  touched={this.state.touched}
                  countries={this.state.countries}
                  updateState={this.updateState}
                  value={this.state.value}
                  autocompleteData={this.state.autocompleteData}
                  capitalize={this.capitalize}
                  calculateTemp={this.calculateTemp}
                />
              </div>
            </div>
          </div>
          <SearchResult
            location={this.state.location}
            temperature={this.state.temperature}
            humidity={this.state.humidity}
            conditions={this.state.conditions}
            error={this.state.error}
            showResult={this.state.showResult}
            isLoading={this.state.isLoading}
            updateState={this.updateState}
          />
        </div>
      </div>
    );
  }
}

export default App;
