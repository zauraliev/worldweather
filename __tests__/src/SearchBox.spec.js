import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import App from "../../src/App";
import SearchBox from "../../src/components/SearchBox";
import SearchBoxInput from "../../src/components/SearchBoxInput";

describe("SearchBox", () => {
  it("getWeather gets weather info with async call", async () => {
    const app = shallow(<App />).instance();
    app.setState({
      touched: {
        city: false,
        country: false
      }
    });

    const wrapper = shallow(
      <SearchBox updateState={app.updateState} touched={app.state.touched} />
    ).instance();

    const preventDefaultSpy = jest.fn();

    const apiKey = "01f0312628c56d7fcadfdee00da7c7a3";
    let apiBaseUrl = "https://api.openweathermap.org/data/2.5/weather?q=";

    const e = {
      target: {
        elements: {
          city: {
            value: "New York"
          },
          country: {
            value: "United States of America"
          }
        }
      }
    };
    wrapper.getWeather({
      preventDefault: preventDefaultSpy,
      target: e.target
    });

    let city = e.target.elements.city.value;
    let country = e.target.elements.country.value;

    wrapper.props.updateState({ isLoading: true });

    expect(app.state.isLoading).toBe(true);

    const callAPI = await fetch(`${apiBaseUrl}${city},${country}&appid=${apiKey}`);

    const data = await callAPI.json();

    expect(data).toBeDefined();

    wrapper.props.updateState({
      location: `${data.name} City, ${data.sys.country}`,
      temperature: app.calculateTemp(data.main.temp),
      humidity: `${data.main.humidity}%`,
      conditions: app.capitalize(data.weather[0].description),
      error: "",
      city: `${city}`,
      country: `${country}`,
      showResult: true,
      isLoading: false
    });

    expect(app.state.showResult).toBe(true);
    expect(app.state.isLoading).toBe(false);
    expect(app.state.location).toBe("New York City, US");
    expect(app.state.temperature).toContain("Â°F");
    expect(app.state.conditions.length).toBeGreaterThan(0);
    expect(app.state.city).toBe("New York");
    expect(app.state.country).toBe("United States of America");

    expect(preventDefaultSpy.mock.calls.length).toBe(1);
  });

  it("updateFormInputs updates form inputs when user is typing", async () => {
    const app = shallow(<App />).instance();
    app.setState({
      touched: {
        city: false,
        country: false
      }
    });
    const wrapper = shallow(
      <SearchBox updateState={app.updateState} touched={app.state.touched} />
    ).instance();
    const preventDefaultSpy = jest.fn();

    const cityInputMock = wrapper.updateFormInputs({
      preventDefault: preventDefaultSpy,
      target: { name: "city", value: "custom value" }
    });
    const cityInput = mount(
      <SearchBoxInput name="city" onChange={cityInputMock} onBlur={cityInputMock} />
    );
    cityInput.find("input").simulate("change");
    expect(preventDefaultSpy.mock.calls.length).toBe(1);

    const countryInputMock = wrapper.updateFormInputs({
      preventDefault: preventDefaultSpy,
      target: { name: "country", value: "custom value" }
    });
    const countryInput = mount(
      <SearchBoxInput name="country" onChange={countryInputMock} onBlur={countryInputMock} />
    );
    countryInput.find("input").simulate("change");
    expect(preventDefaultSpy.mock.calls.length).toBe(2);
  });

  it("city & country are empty test", async () => {
    const app = shallow(<App />).instance();
    app.setState({
      touched: {
        city: false,
        country: false
      }
    });
    const wrapper = shallow(
      <SearchBox updateState={app.updateState} touched={app.state.touched} />
    ).instance();

    const preventDefaultSpy = jest.fn();

    const apiKey = "01f0312628c56d7fcadfdee00da7c7a3";
    let apiBaseUrl = "https://api.openweathermap.org/data/2.5/weather?q=";
    const e = {
      target: {
        elements: {
          city: {
            value: ""
          },
          country: {
            value: ""
          }
        }
      }
    };

    wrapper.getWeather({
      preventDefault: preventDefaultSpy,
      target: e.target
    });

    let city = e.target.elements.city.value;
    let country = e.target.elements.country.value;

    let callAPIMock = { ok: false };
    callAPIMock = await fetch(`${apiBaseUrl}${city},${country}&appid=${apiKey}`);

    expect(app.state.touched.city).toBe(true);
    expect(app.state.touched.country).toBe(true);
  });

  it("renders correctly", () => {
    const app = shallow(<App />).instance();
    app.setState({
      touched: {
        city: false,
        country: false
      }
    });
    const tree = renderer
      .create(<SearchBox updateState={app.updateState} touched={app.state.touched} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
