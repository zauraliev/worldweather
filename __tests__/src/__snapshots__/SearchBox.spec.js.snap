// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders correctly 1`] = `
<form
  className="needs-validation "
  id="weatherForm"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      id="city"
      name="city"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="City"
      required={false}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      id="country"
      name="country"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Country"
      required={false}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <button
      className="btn btn-main btn-md btn-block"
    >
      <span
        className="btn-text"
      >
        Get Weather
      </span>
    </button>
  </div>
</form>
`;
